Nodejs/Setup/InstallerTelemetryAction/InstallerTelemetryAction.cs

```diff
@@ -0,0 +1,313 @@
+                        {
+                            "baseData", new Dictionary < string, object > ()
+                            {
+                                {
+                                    "ver", AIVersion
+                                },
+                                {
+                                    "name", "NtvsInstallerTelemetry"
+                                },
+                                {
+                                    "properties", new Dictionary < string, string > ()
+                                    {
+                                        {
+                                            "InstallStatus", "Error"
+                                        },
+                                        {
+                                            "IsNtvsInstalled", isInstalled.ToString()
+                                        },
+                                        {
+                                            "CurrentState", currentState
+                                        },
+                                        {
+                                            "RequestState", requestState
+                                        },
+                                        {
+                                            "NtvsVersion", session["NtvsVersion"]
+                                        },
+                                        {
+                                            "VSVersion", session["VSVersion"]
+                                        },
+                                        {
+                                            "MsiVersion", session["MsiVersion"]
+                                        },
+                                        {
+                                            "TimeTakenInSeconds", installTime.TotalSeconds.ToString(CultureInfo.InvariantCulture)
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            };
+
+            string jsonString = (new JavaScriptSerializer()).Serialize(data);
+            using (WebClient client = new WebClient())
+            {
+                string response = client.UploadString("https://dc.services.visualstudio.com/v2/track", jsonString);
+                session.Log(response);
+            }
+            session.Log("End Telemetry Log");
+            return ActionResult.Success;
+        }
+
+        [CustomAction]
+        public static ActionResult LogInstallCancelResult(Session session)
+        {
+            TimeSpan installTime = DateTime.Now - DateTime.Parse(session["InstallStartTime"]);
+            string AIVersion = "2";
+
+            session.Log("Begin Telemetry Log");
+            bool isInstalled = session.EvaluateCondition("Installed");
+            string currentState = "";
+            string requestState = "";

```

Comment:
you reassign these later, so there's no need to instantiate them here - it's okay for them to be null.